src/java/org/apache/cassandra/locator/Locator.java

```diff
@@ -0,0 +1,148 @@
+        private VersionedLocation(Epoch epoch, Location location)
+        {
+            this.epoch = epoch;
+            this.location = location;
+        }
+    }
+
+    public Locator(InetAddressAndPort localEndpoint, InitialLocationProvider provider)
+    {
+        this.localEndpoint = localEndpoint;
+        this.initializationLocation = provider.initialLocation();
+        this.local = new AtomicReference<>(new VersionedLocation(Epoch.EMPTY, Location.UNKNOWN));
+    }
+
+    public static Locator forClients()
+    {
+        return new Locator(FBUtilities.getBroadcastAddressAndPort(), () -> Location.UNKNOWN);
+    }
+
+    @VisibleForTesting
+    public void resetState()
+    {
+        state.set(State.INITIAL);
+    }
+
+    public void onInitialized()
+    {
+        logger.info("Node is initialized, moving snitch adapter PREREGISTED state");
+        if (!state.compareAndSet(State.INITIAL, State.UNREGISTERED))
+            throw new IllegalStateException(String.format("Cannot move snitch adapter to UNREGISTERED state (%s)", state.get()));
+    }
+
+    public void onRegistration()
+    {
+        // This may have been done already if the metadata log replay at start up included our registration
+        State current = state.get();
+        if (current == State.REGISTERED)
+            return;
+
+        logger.info("Node is registered, interrupting any previously established peer connections");
+        state.getAndSet(State.REGISTERED);
+        MessagingService.instance().channelManagers.keySet().forEach(MessagingService.instance()::interruptOutbound);
+    }
+
+    public void onPeerRegistration(InetAddressAndPort endpoint)
+    {
+        logger.info("Peer has registered, interrupting any previously established connections");
+        MessagingService.instance().interruptOutbound(endpoint);
+    }
+
+    public Location location(InetAddressAndPort endpoint)
+    {
+        switch (state.get())
+        {
+            case INITIAL:
+                return endpoint.equals(localEndpoint) ? initializationLocation : Location.UNKNOWN;
+            case UNREGISTERED:
+                return endpoint.equals(localEndpoint) ? initializationLocation : fromClusterMetadata(endpoint);
+            default:
+                return fromClusterMetadata(endpoint);
+        }
+    }
+
+    public Location local()
+    {
+        switch (state.get())
+        {
+            case INITIAL:
+            case UNREGISTERED:
+                return initializationLocation;
+            default:
+                // For now, local location is immutable and once registered with cluster metadata, it cannot be
+                // changed. Revisit this if that assumption changes.
+                VersionedLocation location = local.get();
+                if (location.epoch.isAfter(Epoch.EMPTY))
+                    return location.location;
+
+                // We should never get to this point
+                // TODO maybe replace with an assertion
```

Comment:
this is the only place we set local, so I think we will always get to this point (once)

Reply:
I think you're right & that comment is just outdated.